File: haskell/2015/day08b/src/Main.hs
module Main ( main ) where

import Data.Int (Int32)
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

encodedLength :: String -> Int
encodedLength line =
    2 + sum (map encodeChar line)
    where
        encodeChar '"' = 2
        encodeChar '\\' = 2
        encodeChar _ = 1

process :: String -> Int32
process content =
    let contentLines = lines content
        codeLens = map length contentLines
        encLens = map encodedLength contentLines
        diffs = zipWith (-) encLens codeLens
    in fromIntegral $ sum diffs

main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            content <- readFile filename
            let result = process content
            putStrLn $ "result = " ++ show result
        _ -> usage progname

File: rust/2015/day08b/src/main.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::process;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn process_file(filename: &str) -> io::Result<u32> {
    let file = File::open(filename)?;
    let reader = io::BufReader::new(file);
    let mut result: u32 = 0;
    for line_result in reader.lines() {
        let line = line_result?;
        let code_len = line.len() as u32;
        let mut enc_len = 0;
        for ch in line.chars() {
            match ch {
                '\\' | '\"' => enc_len += 2,
                _ => enc_len += 1,
            }
        }
        result += 2 + (enc_len - code_len);
    }
    Ok(result)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}
